js/compiler/compiler.go

```diff
@@ -44,103 +44,114 @@ var (
 		"highlightCode": false,
 	}
 
-	compilerInstance *Compiler
-	once             sync.Once
+	once sync.Once // nolint:gochecknoglobals
+	babl *babel    // nolint:gochecknoglobals
 )
 
-// A Compiler uses Babel to compile ES6 code into something ES5-compatible.
-type Compiler struct {
-	vm *goja.Runtime
+// CompatibilityMode specifies the JS compatibility mode
+// nolint:lll
+//go:generate enumer -type=CompatibilityMode -transform=snake -trimprefix CompatibilityMode -output compatibility_mode_gen.go
+type CompatibilityMode uint8
 
-	// JS pointers.
-	this      goja.Value
-	transform goja.Callable
-	mutex     sync.Mutex //TODO: cache goja.CompileAST() in an init() function?
-}
+const (
+	// CompatibilityModeExtended achieves ES6+ compatibility with Babel and core.js
+	CompatibilityModeExtended CompatibilityMode = iota + 1
+	// CompatibilityModeBase is standard goja ES5.1+
+	CompatibilityModeBase
+)
 
-// Constructs a new compiler.
-func New() (*Compiler, error) {
-	var err error
-	once.Do(func() {
-		compilerInstance, err = new()
-	})
+// A Compiler compiles JavaScript source code (ES5.1 or ES6) into a goja.Program
+type Compiler struct{}
 
-	return compilerInstance, err
+// New returns a new Compiler
+func New() *Compiler {
+	return &Compiler{}
 }
 
-func new() (*Compiler, error) {
-	conf := rice.Config{
-		LocateOrder: []rice.LocateMethod{rice.LocateEmbedded},
+// Transform the given code into ES5, while synchronizing to ensure only a single
+// babel instance is in use.
+func (c *Compiler) Transform(src, filename string) (code string, srcmap SourceMap, err error) {
+	var b *babel
+	if b, err = newBabel(); err != nil {
+		return
 	}
 
-	babelSrc := conf.MustFindBox("lib").MustString("babel.min.js")
+	b.mutex.Lock()
+	defer b.mutex.Unlock()
```

Comment:
Hhmm possibly... Can you substantiate your argument? :)

I feel like the mutex is only needed from the external perspective to `babel`, i.e. `babel` itself shouldn't care how it's used (concurrently or not), that's more up to `Compiler` because it wants to control `babel`s execution.

But I also see the point that `babel` should guarantee thread-safety to its users, so `Compiler` shouldn't care about the `mutex`, which is a private field anyway, so this belongs in `babel.Transform`.

I'm now leaning towards this second point, but let me know if this aligns with your reasoning.

Reply:
My reasoning is that if calling method `A` isn't safe in a concurrent context this should be documented. So now you either document that method `A` is unsafe to call concurrently and then make certain in all places where it's called to not ... actually call it concurrently or well you make it safe for concurrent calls :). (The second one is always a much better option IMO)

In the particular case you are even using the internal to babel Mutex(as you've said yourself). even if you weren't I would argue that you should just make it internal to babel, if that was the only place where an internal to `Compiler` mutex was used.

Reply:
Fixed in 01ebd7b8.